working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,720 @@
+# Extension Types
+
+Author: eernst@google.com
+
+Status: Draft
+
+## Change Log
+
+2021.02.15
+  - Initial version, based on 
+    [language issue 1426](https://github.com/dart-lang/language/issues/1426).
+
+## Introduction
+
+This specifies support for extension types in Dart as an enhancement of the
+extension methods mechanism that Dart has supported since version 2.6.
+
+An _extension type_ is a zero-cost abstraction mechanism that allows
+developers to replace the set of available operations on a given object
+(that is, the instance members of its type) by a different set of
+operations (the members declared by the given extension type).
+
+The point is that the extension type allows for a convenient and safe
+treatment of a given object `o` (and objects reachable from `o`) for a
+specialized purpose or view. It is in particular aimed at the situation
+where that purpose or view requires a certain discipline in the use of
+`o`'s instance methods: We may call certain methods, but only in specific
+ways, and other methods should not be called at all. This kind of added
+discipline can be enforced by accessing `o` typed as an extension type,
+rather than typed as its run-time type `R` or some supertype of `R` (which
+is what we normally do).
+
+An important application would be generated extension types, handling the
+navigation of dynamic object trees. For instance, they could be JSON
+values, modeled using `num`, `bool`, `String`, `List<dynamic>`, and
+`Map<String, dynamic>`.
+
+Without extension types, the JSON value would most likely be handled with
+static type `dynamic`, and all operations on it would be unsafe. If the
+JSON value is assumed to satisfy a specific schema then it would be
+possible to reason about this dynamic code and navigate the tree correctly
+according to the schema. But there is no encapsulation, and the code where
+this kind of careful reasoning is required may be fragmented into many
+different locations.
```

Comment:
Makes sense.

A static type is an abstraction, a *protocol*, for accessing the underlying data.
If you stay within the abstraction, you can maintain certain invariants that you cannot guarantee for direct access to the underlying data. 
(And constructors are there to ensure that the data satisfies the invariant before allowing it to b seen as the abstraction).

Extension types are leaky abstractions. They cannot enforce encapsulation like classes can, because you can always access the underlying data directly if you want to. This differs from how a class can *prevent* access to its underlying implementation.